packages/react/src/xcss/xcss-prop.ts

```diff
@@ -0,0 +1,50 @@
+import type * as CSS from 'csstype';
+
+import type { Pseudos as CSSPseudos } from './pseudos';
+
+export type CSSRuleDefinition = CSSProperties & CSSPseudoRule;
+
+type CSSProperties = Readonly<CSS.PropertiesFallback<number | string>>;
+
+type CSSPseudoRule = { [Q in CSSPseudos]?: CSSProperties };
+
+declare const __classref: unique symbol;
+export type CompiledStyleClassReference = { [__classref]: true };
+
+type XCSSItem<TStyleDecl extends keyof CSSProperties> = {
+  [Q in keyof CSSProperties]: Q extends TStyleDecl ? CompiledStyleClassReference : never;
+};
+
+type XCSSPseudos<K extends keyof CSSProperties, TPseudos extends CSSPseudos> = {
+  [Q in CSSPseudos]?: Q extends TPseudos ? XCSSItem<K> : never;
+};
+
+/**
+ * __xcss prop__
+ *
+ * Controls the type of `xcss` supplied to a component allowing for
+ * more restrictive typing.
+ *
+ * @example
+ * ```tsx
+ * type Props = {
+ *   xcss: XCSSProp<'color', '&:hover'>
+ * }
+ *
+ * const Component = ({ xcss }: Props) => {
+ *
+ * }
+ *
+ * // Valid usage
+ * <Component xcss={{ color: 'red', '&:hover': { color: 'blue' } }} />
+ *
+ * // Invalid usage
+ * <Component xcss={{ backgroundColor: 'red' }} />
+ * ```
+ *
+ */
+export type XCSSProp<AllowedProperties extends keyof CSSProperties, TPseudos extends CSSPseudos> =
+  | (XCSSItem<AllowedProperties> & XCSSPseudos<AllowedProperties, TPseudos>)
```

Comment:
Just to confirm that there won't be need for atrules overrides?

Reply:
No I need to add some I was actually going to merge your approach in `cssMap` so they speak the same language.

Reply:
Sounds good 👍

Reply:
Also just wanted to confirm, the plan is that it will be impossible by design to specify more complex selectors (e.g. `:not(:active)`) and child selectors (e.g. `div span`) using `xcss` right?

cssMap currently does support these through the `selectors` object (https://compiledcssinjs.com/docs/api-cssmap#nested-selectors-and-advanced-selectors) but this can be forbidden through linting rules